samples/wifi/provisioning/common/wifi_prov_ble.c

```diff
@@ -0,0 +1,407 @@
+
+	/* If no config, mark it as unprovisioned. */
+	if (!wifi_has_config()) {
+		prov_svc_data[ADV_DATA_FLAG_IDX] &= ~ADV_DATA_FLAG_PROV_STATUS_BIT;
+	} else {
+		prov_svc_data[ADV_DATA_FLAG_IDX] |= ADV_DATA_FLAG_PROV_STATUS_BIT;
+	}
+
+	rc = net_mgmt(NET_REQUEST_WIFI_IFACE_STATUS, iface, &status,
+				sizeof(struct wifi_iface_status));
+	/* If WiFi is not connected or error occurs, mark it as not connected. */
+	if ((rc != 0) || (status.state < WIFI_STATE_ASSOCIATED)) {
+		prov_svc_data[ADV_DATA_FLAG_IDX] &= ~ADV_DATA_FLAG_CONN_STATUS_BIT;
+		prov_svc_data[ADV_DATA_RSSI_IDX] = INT8_MIN;
+	} else {
+		/* WiFi is connected. */
+		prov_svc_data[ADV_DATA_FLAG_IDX] |= ADV_DATA_FLAG_CONN_STATUS_BIT;
+		/* Currently cannot retrieve RSSI. Use a dummy number. */
+		prov_svc_data[ADV_DATA_RSSI_IDX] = status.rssi;
+	}
+
+	rc = bt_le_adv_update_data(ad, ARRAY_SIZE(ad), sd, ARRAY_SIZE(sd));
+	if (rc != 0) {
+		LOG_DBG("Cannot update advertisement data, err = %d", rc);
+	}
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		LOG_WRN("BT Connection failed (err 0x%02x).", err);
+	} else {
+		LOG_INF("BT Connected.");
+		default_conn = conn;
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	LOG_INF("BT Disconnected (reason 0x%02x).", reason);
+	bt_unpair(BT_ID_DEFAULT, NULL);
+	default_conn = NULL;
+	k_work_reschedule(&update_adv_param_work, K_NO_WAIT);
+}
+
+static void identity_resolved(struct bt_conn *conn, const bt_addr_le_t *rpa,
+				  const bt_addr_le_t *identity)
+{
+	char addr_identity[BT_ADDR_LE_STR_LEN];
+	char addr_rpa[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(identity, addr_identity, sizeof(addr_identity));
+	bt_addr_le_to_str(rpa, addr_rpa, sizeof(addr_rpa));
+
+	LOG_INF("BT Identity resolved %s -> %s.", addr_rpa, addr_identity);
+}
+
+static void security_changed(struct bt_conn *conn, bt_security_t level,
+				 enum bt_security_err err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (!err) {
+		LOG_INF("BT Security changed: %s level %u.", addr, level);
+	} else {
+		LOG_WRN("BT Security failed: %s level %u err %d.", addr, level,
+			   err);
+	}
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.identity_resolved = identity_resolved,
+	.security_changed = security_changed,
+};
+
+static void auth_cancel(struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_INF("BT Pairing cancelled: %s.", addr);
+}
+
+static struct bt_conn_auth_cb auth_cb_display = {
+	.cancel = auth_cancel,
+};
+
+static void pairing_complete(struct bt_conn *conn, bool bonded)
+{
+	LOG_INF("BT Pairing Complete.");
+}
+
+static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason)
+{
+	LOG_INF("BT Pairing Failed (%d). Disconnecting.", reason);
+	bt_conn_disconnect(conn, BT_HCI_ERR_AUTH_FAIL);
```

Comment:
Ideally when the pairing fails, the app side (Android or iOS) should disconnect the BT. But we observe a long delay between the app drops the connection and the OS releases the connnection. So we do it on device side to enforce the disconnection. We will investigate if this can be solved on app side, and then this `bt_conn_disconnect` can be removed.